File: activity/20200805/junhaolin/1219.Path_with_Maximum_Gold.cpp
class Solution {
public:
    vector<vector<int>> grid;
    int maxSum;
    int dx [4] = {-1,0,1,0}, dy [4] = {0,-1,0,1};
    
    int getMaximumGold(vector<vector<int>>& grid) {
        this->grid = grid;
        this->maxSum = 0;
        
        for (int i=0; i<grid.size(); i++) 
            for (int j=0; j<grid[0].size(); j++) 
                traverse(0, i, j);
        
        return maxSum;
    }
    
    void traverse(int sum, int x, int y) {
        if (grid[x][y] > 0) {
            int value = grid[x][y];
            sum += value;
            grid[x][y] = 0;
            
            maxSum = max(maxSum, sum);
            
            for (int i=0; i<4; i++) {
                int X = x+dx[i], Y = y+dy[i];
                if (0 <= X && X < grid.size() && 0 <= Y && Y < grid[0].size()) {
                    traverse(sum, X, Y);
                }
            }
            
            grid[x][y] = value;
        }
    }
};

File: activity/20200805/ninezero/1219_path-with-maximum-gold.go
func getMaximumGold(grid [][]int) int {
	// 边界判断
	if grid == nil || len(grid) == 0 {
		return 0
	}

	var res float64

	// 循环每一个点，找到最大值
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[0]); j++ {
			res = math.Max(res, dfs(grid, i, j))
		}
	}

	return int(res)
}

func dfs(grid [][]int, x, y int) float64 {
	// 结束条件
	if x < 0 || y < 0 || x >= len(grid) || y >= len(grid[0]) || grid[x][y] == 0 {
		return 0
	}

	// 保存当前节点值，回溯算法
	temp := grid[x][y]

	// 递归查找上下左右四个方向
	grid[x][y] = 0
	up := dfs(grid, x-1, y)
	down := dfs(grid, x+1, y)
	left := dfs(grid, x, y-1)
	right := dfs(grid, x, y+1)

	// 上下左右四个方向的最大值
	var res float64
	res = math.Max(res, left)
	res = math.Max(res, right)
	res = math.Max(res, up)
	res = math.Max(res, down)

	grid[x][y] = temp

	// 加上当前grid的值
	return res + float64(grid[x][y])
}
